File: back/example/worlds/billardBall.go
package exampleWorlds

import (
	"fmt"
	"math"
	"root/common"
	"root/world"
)

func BillardBall() *world.World {
	w := world.NewWorld("billardBall", 0.01)

	w.Objects = append(w.Objects, common.NewObject(
		"table", "#654321",
		common.NewHollowRectFrame(2.8, 1.4, 0.05, 0.1),
		1000,
		common.Vector3{X: 0, Y: 0.1 / 2, Z: 0},
		common.Vector3{},
	))

	// Ball parameters
	radius := float32(0.057)
	spacing := float32(0.01)
	startX := float32(0.7)
	triangleRows := 4
	ballColor := "#ffffff"
	ballDensity := float32(170) // 170kg/m^3 (arbitrary, just for mass)

	// Place triangle of balls
	for row := 0; row < triangleRows; row++ {
		for col := 0; col <= row; col++ {
			x := startX + float32(row)*(radius*2+spacing)*float32(math.Sin(math.Pi/3))
			z := (float32(col) - float32(row)/2.0) * ((radius * 2) + spacing)
			w.Objects = append(w.Objects, common.NewObject(
				fmt.Sprintf("ball_%d_%d", row, col), ballColor,
				common.NewSphere(radius, 0),
				ballDensity,
				common.Vector3{X: x, Y: radius, Z: z},
				common.Vector3{},
			))
		}
	}

	cue := common.NewObject(
		"cue", "#ffff00",
		common.NewSphere(radius, 0),
		ballDensity,
		common.Vector3{X: -1.0, Y: radius, Z: 0.0},
		common.Vector3{X: 2.0, Y: 0, Z: 0},
	)
	w.Objects = append(w.Objects, cue)

	return w
}

File: back/example_worlds/billard_ball.py
"""
Billard ball simulation (matches Go version exactly).
"""
import torch
import config

from world import World
from obj import Obj
from shapes import sphere, hollow_rect_frame
from recording import Recording


def simulate_billard_ball():
    print(f"Device: {config.device}")

    # Create world
    world = World('billardBall', time_step=0.01)

    # Table frame
    table = Obj(
        'table', '#654321',
        hollow_rect_frame(280.0, 140.0, 5.0, 10.0),
        1000,
        torch.tensor([0., 10.0 / 2, 0.], device=config.device),
        torch.tensor([0., 0., 0.], device=config.device)
    )
    world.add_object(table)

    # Ball parameters
    radius = 5.7
    spacing = 0.0
    ball_density = 170
    balls_per_row = 5
    row_count = 3
    row_separation = 20.0
    start_x = -70.0
    start_z = -20.0
    move_speed = 200.0

    for row in range(row_count):
        moving_balls = row + 1
        stationary_balls = balls_per_row - moving_balls
        z = start_z + row * row_separation

        # Place stationary balls
        for i in range(stationary_balls):
            x = start_x + i * (radius * 2 + spacing)
            ball = Obj(
                f'row{row}_stat{i}', '#ffffff',
                sphere(radius, 0),
                ball_density,
                torch.tensor([x, radius, z], device=config.device),
                torch.tensor([0., 0., 0.], device=config.device)
            )
            world.add_object(ball)

        # Place moving balls
        for i in range(moving_balls):
            x = start_x + (stationary_balls + i) * (radius * 2 + spacing)
            ball = Obj(
                f'row{row}_mov{i}', '#ffff00',
                sphere(radius, 0),
                ball_density,
                torch.tensor([x, radius, z], device=config.device),
                torch.tensor([-move_speed, 0., 0.], device=config.device)
            )
            world.add_object(ball)

    print(f"World has {len(world.objects)} objects")

    # Create and run recording (matches Go: 0.01 seconds per frame, 10 seconds total)
    recording = Recording('billardBall', world, seconds_per_frame=0.01, end_seconds=10)
    recording.simulate()


if __name__ == '__main__':
    simulate_billard_ball()
